File: src/bmsx/vm/font.ts
import { BFont, GlyphMap } from '../core/font';

export type VMFontVariant = 'msx' | 'tiny';

type VMFontPreset = {
	prefix: string;
	tabDirtyMarkerAssetId: string;
	buildCharMap(): GlyphMap;
};

export const DEFAULT_VM_FONT_VARIANT: VMFontVariant = 'msx';

const FONT_PRESETS: Record<VMFontVariant, VMFontPreset> = {
	msx: {
		prefix: 'msx_6b_font',
		tabDirtyMarkerAssetId: 'msx_6b_font_ctrl_bel',
		buildCharMap(): GlyphMap {
			return buildMsxCharMap('msx_6b_font');
		},
	},
	tiny: {
		prefix: 'tiny_3b_font',
		tabDirtyMarkerAssetId: 'tiny_3b_font_ctrl_bel',
		buildCharMap(): GlyphMap {
			return buildTinyCharMap('tiny_3b_font');
		},
	},
} as const;

export function getVMFontPreset(variant: VMFontVariant): VMFontPreset {
	return FONT_PRESETS[variant];
}

function buildMsxCharMap(prefix: string): GlyphMap {
	const withPrefix = (suffix: string): string => `${prefix}_${suffix}`;
	const map: GlyphMap = {
		' ': withPrefix('space'),
		'!': withPrefix('exclamation'),
		'"': withPrefix('code_0x22'),
		'#': withPrefix('code_0x23'),
		'$': withPrefix('code_0x24'),
		'%': withPrefix('percent'),
		'&': withPrefix('code_0x26'),
		'\'': withPrefix('apostroph'),
		'(': withPrefix('code_0x28'),
		')': withPrefix('code_0x29'),
		'*': withPrefix('code_0x2a'),
		'+': withPrefix('code_0x2b'),
		',': withPrefix('comma'),
		'-': withPrefix('streep'),
		'.': withPrefix('dot'),
		'/': withPrefix('slash'),
		':': withPrefix('colon'),
		';': withPrefix('code_0x3b'),
		'<': withPrefix('code_0x3c'),
		'=': withPrefix('code_0x3d'),
		'>': withPrefix('code_0x3e'),
		'?': withPrefix('question'),
		'@': withPrefix('at_sign'),
		'[': withPrefix('code_0x5b'),
		'\\': withPrefix('code_0x5c'),
		']': withPrefix('code_0x5d'),
		'^': withPrefix('code_0x5e'),
		'_': withPrefix('line'),
		'`': withPrefix('code_0x60'),
		'{': withPrefix('code_0x7b'),
		'|': withPrefix('code_0x7c'),
		'}': withPrefix('code_0x7d'),
		'~': withPrefix('code_0x7e'),
		'•': withPrefix('ctrl_bel'),
		'¡': withPrefix('code_0x80'),
	};
	for (let i = 0; i < 10; i += 1) {
		const digit = String.fromCharCode(48 + i);
		map[digit] = withPrefix(digit);
	}
	const lowercase = 'abcdefghijklmnopqrstuvwxyz';
	for (let i = 0; i < lowercase.length; i += 1) {
		const ch = lowercase.charAt(i);
		map[ch] = withPrefix(`low_${ch}`);
	}
	const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
	for (let i = 0; i < uppercase.length; i += 1) {
		const upper = uppercase.charAt(i);
		const lower = upper.toLowerCase();
		map[upper] = withPrefix(lower);
	}
	return map;
}

function buildTinyCharMap(prefix: string): GlyphMap {
	const withPrefix = (suffix: string): string => `${prefix}_${suffix}`;
	const map: GlyphMap = {
		' ': withPrefix('space'),
		'!': withPrefix('exclamation'),
		'@': withPrefix('at_sign'),
		'#': withPrefix('hash'),
		'$': withPrefix('dollar'),
		'%': withPrefix('percent'),
		'&': withPrefix('ampersand'),
		'"': withPrefix('quote'),
		'\'': withPrefix('apostroph'),
		'(': withPrefix('parenopen'),
		')': withPrefix('parenclose'),
		'*': withPrefix('asterisk'),
		'+': withPrefix('plus'),
		',': withPrefix('comma'),
		'-': withPrefix('streep'),
		'.': withPrefix('dot'),
		'/': withPrefix('slash'),
		':': withPrefix('colon'),
		';': withPrefix('semicolon'),
		'~': withPrefix('tilde'),
		'<': withPrefix('lessthan'),
		'=': withPrefix('equals'),
		'>': withPrefix('greaterthan'),
		'?': withPrefix('question'),
		'[': withPrefix('bracketopen'),
		'\\': withPrefix('backslash'),
		']': withPrefix('bracketclose'),
		'^': withPrefix('caret'),
		'_': withPrefix('line'),
		'`': withPrefix('backtick'),
		'{': withPrefix('braceopen'),
		'|': withPrefix('pipe'),
		'}': withPrefix('braceclose'),
		'•': withPrefix('bullet'),
		'¡': withPrefix('inverted_exclamation'),
		'¤': withPrefix('flower'),
		'¦': withPrefix('brokenbar'),
		'§': withPrefix('section'),
		'£': withPrefix('pound'),
		'¥': withPrefix('yen'),
		'€': withPrefix('euro'),
		'µ': withPrefix('euler'),
		'ĳ': withPrefix('low_ij'),
		'Ĳ': withPrefix('ij'),
	};
	for (let i = 0; i < 10; i += 1) {
		const digit = String.fromCharCode(48 + i);
		map[digit] = withPrefix(digit);
	}
	const lowercase = 'abcdefghijklmnopqrstuvwxyz';
	for (let i = 0; i < lowercase.length; i += 1) {
		const ch = lowercase.charAt(i);
		map[ch] = withPrefix(`low_${ch}`);
	}
	const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
	for (let i = 0; i < uppercase.length; i += 1) {
		const upper = uppercase.charAt(i);
		const lower = upper.toLowerCase();
		map[upper] = withPrefix(lower);
	}
	return map;
}

export class VMFont extends BFont {
	protected readonly preset: VMFontPreset;
	protected readonly variant: VMFontVariant;

	constructor(config?: { variant?: VMFontVariant }) {
		const variant = config?.variant ?? DEFAULT_VM_FONT_VARIANT;
		const preset = getVMFontPreset(variant);
		super(preset.buildCharMap());
		this.variant = variant;
		this.preset = preset;
		// this.resetLetterMap();
	}

	protected resetLetterMap(): void {
		const target = this.letter_to_img;
		const keys = Object.keys(target);
		for (let i = 0; i < keys.length; i++) {
			delete target[keys[i]];
		}
		const map = this.preset.buildCharMap();
		const entries = Object.keys(map);
		for (let i = 0; i < entries.length; i++) {
			const ch = entries[i];
			target[ch] = map[ch];
		}
	}

}

File: src/bmsx_cpp/vm/font.cpp
/*
 * font.cpp - VM font variants
 */

#include "font.h"

namespace bmsx {
namespace {

GlyphMap buildMsxCharMap() {
    const std::string prefix = "msx_6b_font";
    auto withPrefix = [&](const std::string& suffix) {
        return prefix + "_" + suffix;
    };

    GlyphMap map;
    map[static_cast<u32>(' ')] = withPrefix("space");
    map[static_cast<u32>('!')] = withPrefix("exclamation");
    map[static_cast<u32>('\"')] = withPrefix("code_0x22");
    map[static_cast<u32>('#')] = withPrefix("code_0x23");
    map[static_cast<u32>('$')] = withPrefix("code_0x24");
    map[static_cast<u32>('%')] = withPrefix("percent");
    map[static_cast<u32>('&')] = withPrefix("code_0x26");
    map[static_cast<u32>('\'')] = withPrefix("apostroph");
    map[static_cast<u32>('(')] = withPrefix("code_0x28");
    map[static_cast<u32>(')')] = withPrefix("code_0x29");
    map[static_cast<u32>('*')] = withPrefix("code_0x2a");
    map[static_cast<u32>('+')] = withPrefix("code_0x2b");
    map[static_cast<u32>(',')] = withPrefix("comma");
    map[static_cast<u32>('-')] = withPrefix("streep");
    map[static_cast<u32>('.')] = withPrefix("dot");
    map[static_cast<u32>('/')] = withPrefix("slash");
    map[static_cast<u32>(':')] = withPrefix("colon");
    map[static_cast<u32>(';')] = withPrefix("code_0x3b");
    map[static_cast<u32>('<')] = withPrefix("code_0x3c");
    map[static_cast<u32>('=')] = withPrefix("code_0x3d");
    map[static_cast<u32>('>')] = withPrefix("code_0x3e");
    map[static_cast<u32>('?')] = withPrefix("question");
    map[static_cast<u32>('@')] = withPrefix("at_sign");
    map[static_cast<u32>('[')] = withPrefix("code_0x5b");
    map[static_cast<u32>('\\')] = withPrefix("code_0x5c");
    map[static_cast<u32>(']')] = withPrefix("code_0x5d");
    map[static_cast<u32>('^')] = withPrefix("code_0x5e");
    map[static_cast<u32>('_')] = withPrefix("line");
    map[static_cast<u32>('`')] = withPrefix("code_0x60");
    map[static_cast<u32>('{')] = withPrefix("code_0x7b");
    map[static_cast<u32>('|')] = withPrefix("code_0x7c");
    map[static_cast<u32>('}')] = withPrefix("code_0x7d");
    map[static_cast<u32>('~')] = withPrefix("code_0x7e");
    map[0x2022] = withPrefix("ctrl_bel");
    map[0x00A1] = withPrefix("code_0x80");

    for (int i = 0; i < 10; ++i) {
        char digit = static_cast<char>('0' + i);
        map[static_cast<u32>(digit)] = withPrefix(std::string(1, digit));
    }

    for (char c = 'a'; c <= 'z'; ++c) {
        map[static_cast<u32>(c)] = withPrefix(std::string("low_") + c);
    }
    for (char c = 'A'; c <= 'Z'; ++c) {
        char lower = static_cast<char>(c - 'A' + 'a');
        map[static_cast<u32>(c)] = withPrefix(std::string(1, lower));
    }

    return map;
}

GlyphMap buildTinyCharMap() {
    const std::string prefix = "tiny_3b_font";
    auto withPrefix = [&](const std::string& suffix) {
        return prefix + "_" + suffix;
    };

    GlyphMap map;
    map[static_cast<u32>(' ')] = withPrefix("space");
    map[static_cast<u32>('!')] = withPrefix("exclamation");
    map[static_cast<u32>('@')] = withPrefix("at_sign");
    map[static_cast<u32>('#')] = withPrefix("hash");
    map[static_cast<u32>('$')] = withPrefix("dollar");
    map[static_cast<u32>('%')] = withPrefix("percent");
    map[static_cast<u32>('&')] = withPrefix("ampersand");
    map[static_cast<u32>('\"')] = withPrefix("quote");
    map[static_cast<u32>('\'')] = withPrefix("apostroph");
    map[static_cast<u32>('(')] = withPrefix("parenopen");
    map[static_cast<u32>(')')] = withPrefix("parenclose");
    map[static_cast<u32>('*')] = withPrefix("asterisk");
    map[static_cast<u32>('+')] = withPrefix("plus");
    map[static_cast<u32>(',')] = withPrefix("comma");
    map[static_cast<u32>('-')] = withPrefix("streep");
    map[static_cast<u32>('.')] = withPrefix("dot");
    map[static_cast<u32>('/')] = withPrefix("slash");
    map[static_cast<u32>(':')] = withPrefix("colon");
    map[static_cast<u32>(';')] = withPrefix("semicolon");
    map[static_cast<u32>('~')] = withPrefix("tilde");
    map[static_cast<u32>('<')] = withPrefix("lessthan");
    map[static_cast<u32>('=')] = withPrefix("equals");
    map[static_cast<u32>('>')] = withPrefix("greaterthan");
    map[static_cast<u32>('?')] = withPrefix("question");
    map[static_cast<u32>('[')] = withPrefix("bracketopen");
    map[static_cast<u32>('\\')] = withPrefix("backslash");
    map[static_cast<u32>(']')] = withPrefix("bracketclose");
    map[static_cast<u32>('^')] = withPrefix("caret");
    map[static_cast<u32>('_')] = withPrefix("line");
    map[static_cast<u32>('`')] = withPrefix("backtick");
    map[static_cast<u32>('{')] = withPrefix("braceopen");
    map[static_cast<u32>('|')] = withPrefix("pipe");
    map[static_cast<u32>('}')] = withPrefix("braceclose");
    map[0x2022] = withPrefix("bullet");
    map[0x00A1] = withPrefix("inverted_exclamation");
    map[0x00A4] = withPrefix("flower");
    map[0x00A6] = withPrefix("brokenbar");
    map[0x00A7] = withPrefix("section");
    map[0x00A3] = withPrefix("pound");
    map[0x00A5] = withPrefix("yen");
    map[0x20AC] = withPrefix("euro");
    map[0x00B5] = withPrefix("euler");
    map[0x0133] = withPrefix("low_ij");
    map[0x0132] = withPrefix("ij");

    for (int i = 0; i < 10; ++i) {
        char digit = static_cast<char>('0' + i);
        map[static_cast<u32>(digit)] = withPrefix(std::string(1, digit));
    }

    for (char c = 'a'; c <= 'z'; ++c) {
        map[static_cast<u32>(c)] = withPrefix(std::string("low_") + c);
    }
    for (char c = 'A'; c <= 'Z'; ++c) {
        char lower = static_cast<char>(c - 'A' + 'a');
        map[static_cast<u32>(c)] = withPrefix(std::string(1, lower));
    }

    return map;
}

} // namespace

VMFont::VMFont(RuntimeAssets& assets, VMFontVariant variant)
    : BFont(assets, variant == VMFontVariant::Tiny ? buildTinyCharMap() : buildMsxCharMap()) {
}

} // namespace bmsx
